samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -0,0 +1,288 @@
+# coding: utf-8
+class FunctionError < StandardError; end
+
+def define_for_evaled_arguments(&implementation)
+  lambda { |arguments, context|
+    evaled_arguments = arguments.map { |argument| eval_zil argument, context }
+    implementation.call(evaled_arguments, context)
+  }
+end
+
+# helper method to check for argument count
+# to be used with rethrowing the error with the `ROUTINE` name in front
+def argc!(arguments, count, cmp_type = :==)
+  case cmp_type
+  when :===
+    raise ArgumentError, "`count` (#{count}) not a `Range` object" unless count.is_a?(Range)
+    raise FunctionError, "has an arity (#{count.begin} ≤ n ≤ #{count.end}), got #{arguments.length}" unless count === arguments.length
+  else
+    raise FunctionError, "has an arity (n #{cmp_type} #{count}), got #{arguments.length}" unless arguments.length.send(cmp_type, count)
+  end
+end
+
+def expect_argument_count!(arguments, expected_count)
+  argc!(arguments, expected_count)
+end
+
+def expect_argument_count_in_range!(arguments, range)
+  argc!(arguments, range, :===)
+end
+
+def expect_minimum_argument_count!(arguments, min_count)
+  argc!(arguments, min_count, :>=)
+end
+
+ZIL_BUILTINS = {}
+
+ZIL_BUILTINS[:LVAL] = define_for_evaled_arguments { |arguments, context|
+  expect_argument_count!(arguments, 1)
+  var_atom = arguments[0]
+  raise FunctionError, "No local value for #{var_atom.inspect}" unless context.locals.key? var_atom
+
+  context.locals[var_atom]
+}
+
+ZIL_BUILTINS[:VALUE] = define_for_evaled_arguments { |arguments, context|
+  expect_argument_count!(arguments, 1)
+  var_atom = arguments[0]
+
+  context.locals[var_atom] || context.globals[var_atom] ||
+    (raise FunctionError, "No local nor global value for #{var_atom.inspect}")
+}
```

Comment:
You're going to want to back this PR out and resubmit with your HEAD at the same spot as the zil-interpreter HEAD.  We should ideally only see the two commits at the end.

Reply:
I took care of that... It was just the base branch of the Pull Request that was not set correctly....